draft-ietf-jsonpath-base.md

```diff
@@ -1317,10 +1327,25 @@ function-expression     = function-name "(" S [function-argument
 function-argument       = comparable / filter-path
 ~~~
 
-Syntactically, a function-expression can occur anywhere where a
-singular-path can occur (exist-expr, comparable).
-A function-expression that employs a function extension that returns a
-nodelist MUST return a singular node to be used in a comparable.
+Any function expressions in a query must be syntactically valid,
+otherwise the JSONPath implementation MUST raise an error
+(see Section {{<synsem-overview}}).
+
+A function expression is syntactically valid if:
+
+* It conforms to the ABNF syntax above.
+* It occurs as a filter-path in an existence test and the function
+is defined to return either a nodelist or a Singular Nodelist.
+* It occurs as a singular-path on either side of a comparison and the function
+is defined to return either a value or a Singular Nodelist.
+* For the function expression and any function expression it contains,
+each argument of the function matches the defined kind of the argument:
+  * An argument of kind value takes a primitive value or a function expression
```

Comment:
This doesn't convert a singular path to a value.
`length($.foo)` would not work?

Reply:
Correct, and that's too limiting. I was trying to avoid the issue a function expecting a value but receiving an empty nodelist from a singular path. Here are some alternatives if we allowed a singular path or a function expression returning a singular nodelist:

* We could coerce an empty nodelist to, for example, `null` at runtime, but we've managed to avoid coercions elsewhere in the spec.
* We could cause the filter not to match for the current item. This would impact the Boolean laws which would no longer be true in general.

Reply:
Another alternative occurred to me. We could introduce a function extension `value_or_else()` which takes a singular nodelist and a default value and returns a value. The result would be the value of the node if the nodelist is non-empty and the default value otherwise. We could then use `length(value_or_null(<singular path>, ""))`. This is probably too long-winded to be practical.

Another alternative would be to define another `length()` function which takes a singular nodelist and which can be defined to return a suitable default (such as 0 or -1) if the nodelist is empty. Not sure we want to get into overloading of function extension names. That could be an issue for IANA.

Reply:
I assume that with `length($.foo)` not working, we imply `$.foo` is taken as an existence test, right?

In the same way as an application needs to distinguish occurences of paths as part of comparisons or just not, they should be able to detect, if a path occurs as a function parameter and might be handed over as it is then. That way the burden/freedom to correctly treat the path lies at the function author (simply taking its value here with `length`).

Have I missed something here?

Reply:
> I assume that with `length($.foo)` not working, we imply `$.foo` is taken as an existence test, right?

Oops.  That was not what I intended -- length wants an array, a map or a text string.
So converting this to a Boolean is not what one wants here.
(But for some other function extension, this might exactly be what is wanted -- however, this could then possibly be forced by `!!$.foo`.)
So maybe we need to distinguish function parameters where the arguments are evaluated in a comparison-like context, and parameters where the arguments are evaluated as an existence test.

Reply:
Yes, `!!$.foo` would be perfectly fine in case of an existence test as parameter.

But ... exactly what prevents `length($.foo)` from working then?

Reply:
Expanding on this functional programming approach, we could define the following types, where `Nodes(l)` is an instance for some nodelist `l` (which could be represented either as a path or as a nodelist):
* `OptionalNodeOrValue = {None, Node(n), Value(v)}`
* `OptionalNode = {None, Node(n)}`
* `OptionalValue = {None, Value(v)}`
* `Value = {Value(v)}`
* `Absent = {None}` (this may not be necessary, but allows, for example, a logging function with no return value)
* `OptionalNodes = {None, Nodes(l)}` (where `None` is equivalent to `Nodes(e)`, where `e` is the empty nodelist)

Each function argument  or return value would have one of the above types. 

The types form a hierarchy (with subtypes indented):
```
OptionalNodeOrValue
    OptionalNode
        Absent
    OptionalValue
        Value
        Absent
    Value
    Absent

OptionalNodes
    OptionalNode
        Absent
```

So, for example, a function with return value of type `Value` could be passed to another function with an argument of type `OptionalValue`.

Note that `OptionalNode` is a subtype of `OptionalNodes` even though the instances of the former are not a subset of the instances of the latter. The `OptionalNode` instance `Node(n)` can be coerced to the `OptionalNodes` instance `Nodes(l)`, where `l` is a nodelist consisting of just the node `n`.

What do people think of this approach?

Reply:
This is far too abstract for me and perhaps for someone else.

From the view of an average user of JSONPath - knowing, there is a function `length`, familiar from C-family of languages - I would expect, that the query

`$[?length(@.something] >= 0]`

simply works regardless of the argument.

How would that query look like using your approach then?

Reply:
I don't think we can avoid the inherent complexity.
Maybe we can slightly simplify the tree.

Defining this in an abstract way is fine for our specification.
Of course, a few examples can make this more accessible.

Reply:
> How would that query look like using your approach then?

The query would be the same. I'm simply describing which function expressions are syntactically valid.

Reply:
> Maybe we can slightly simplify the tree.

Note the type hierarchy (if that's the right word) isn't a tree since there are nodes such as `Value` with multiple parents.

Reply:
> Of course, a few examples can make this more accessible.

For starters, here are some proposed signatures of the functions that are already in the spec:
* `length(OptionalNodeOrValue) -> OptionalValue`
* `count(OptionalNodes) -> Value`
* `match(Value) -> OptionalValue`
* `search(Value) -> OptionalValue`

For all these that return `OptionalValue`, `None` (i.e. an empty nodelist) is used in error cases (e.g. an empty nodelist argument to `length` or a non-string argument of `match` or `search`).

Reply:
This makes it a bit clearer ... thanks. Also my example 

`$[?length(@.something] >= 0]`

seems to be in harmony with our comparison rules now.